README.md
# Spok

By [Magnetis](https://magnetis.com.br)

[![Build Status](https://travis-ci.org/magnetis/spok.svg?branch=master)](https://travis-ci.org/magnetis/spok)

Spok is a tool for dealing with workdays and restdays in an easy way.
It also provides functionalities for working with periods of dates.

## Installation

Add this line to your application's Gemfile:

    gem 'spok'

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install spok

## Usage

### Dealing with workdays

You can use the `Spok::Workday` module to check if a date is either a restday or a workday:

```ruby
require 'spok'

Spok::Workday.workday?(Date.new(2012, 12, 24))
# => true
```

Spok also supports different calendars for checking if a date is a workday or a
restday

```ruby
require 'spok'

Spok::Workday.workday?(Date.new(2012, 12, 24), calendar: :bovespa)
# => false
```

If `calendar` is not informed, it will default to the `:brasil` calendar.

The available calendars are defined with `.yml` files [here](lib/spok/config/).

### Working with periods of dates

Objects from the `Spok` class can be used to work with periods of dates. You can
initialize a new object passing the start and end date for your period to the
`new` method:

```ruby
require 'spok'

period = Spok.new(Date.new(2018, 9, 14), Date.new(2018, 10, 14))
another_period = Spok.new(Date.today, Date.today + 10.days)
```

You can also use the `Spok.parse` method to initialize a new object using a string:

```ruby
require 'spok'

parsed_period = Spok.parse('20180901-20180903')
```

With the object, you can check for all the workdays in that period of dates:

```ruby
require 'spok'

period = Spok.new(Date.new(2018, 10, 10), Date.new(2018, 10, 14))
period.workdays
# => [Wed, 10 Oct 2018, Thu, 11 Oct 2018]
```

## Documentation

The complete documentation in the RDoc format is available here:

https://www.rubydoc.info/gems/spok

## Contributing

Please read our [Contributing](CONTRIBUTING.md) guide for more information.

## License

This project is licensed under the terms of the Apache 2. Copyright 2014-2018 Magnetis http://www.magnetis.com.br
See the [LICENSE](LICENSE.txt) file for license rights and limitations (Apache 2).


CHANGELOG.md
# Change log

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [1.0.0] - 2018-09-14
- First public release.


spec/spec_helper.rb
# This file was generated by the `rspec --init` command. Conventionally, all
# specs live under a `spec` directory, which RSpec adds to the `$LOAD_PATH`.
# Require this file using `require "spec_helper"` to ensure that it is only
# loaded once.
#
# See http://rubydoc.info/gems/rspec-core/RSpec/Core/Configuration
RSpec.configure do |config|
  config.run_all_when_everything_filtered = true
  config.filter_run :focus

  # Run specs in random order to surface order dependencies. If you find an
  # order dependency and want to debug it, you can fix the order by providing
  # the seed, which is printed after each run.
  #     --seed 1234
  config.order = 'random'
  config.expect_with(:rspec) { |c| c.syntax = %i(expect) }
end


spec/lib/spok_spec.rb
require 'spec_helper'
require 'spok'

describe Spok do
  let(:sunday) { Date.new(2012, 1, 1) }
  let(:monday) { Date.new(2012, 1, 2) }
  let(:tuesday) { Date.new(2012, 1, 3) }

  subject(:spok) { described_class.new(sunday, tuesday) }

  describe '.parse' do
    it { expect(described_class.parse('20120101-20120103')).to eq(spok) }

    context 'when the input string is wrong' do
      it { expect(described_class.parse('20120101')).to be_nil }
      it { expect(described_class.parse(nil)).to be_nil }
      it { expect(described_class.parse('')).to be_nil }
    end
  end

  describe '#initialize' do
    it 'does not raise error when start date is before or equals to the end date' do
      expect { subject }.not_to raise_error
      expect { described_class.new(sunday, sunday) }.not_to raise_error
    end

    it 'raises exception when the start date is nil' do
      expect { described_class.new(nil, sunday) }.to raise_error(ArgumentError)
    end

    it 'raises exception when the end date is nil' do
      expect { described_class.new(tuesday, nil) }.to raise_error(ArgumentError)
    end

    it 'raises exception when the start date is greater than the end date' do
      expect { described_class.new(tuesday, sunday) }.to raise_error(ArgumentError)
    end
  end

  describe '#date_as_integer' do
    it 'represents the start date as integer' do
      expect(subject.start_date_as_integer).to eq(20120101)
    end

    it 'represents the end date as integer' do
      expect(subject.end_date_as_integer).to eq(20120103)
    end
  end

  describe '#date_as_string' do
    it 'represents the start date as string' do
      expect(subject.start_date_as_string).to eq('20120101')
    end

    it 'represents the end date as string' do
      expect(subject.end_date_as_string).to eq('20120103')
    end
  end

  describe '#to_s' do
    it 'successfully converts Spok to String' do
      expect(spok.to_s).to eq('20120101-20120103')
    end
  end

  describe '#workdays' do
    subject(:spok) { described_class.new(sunday, wednesday) }

    let(:sunday) { Date.new(2013, 7, 7) }
    let(:monday) { Date.new(2013, 7, 8) }
    let(:bovespa_holiday) { Date.new(2013, 7, 9) }
    let(:wednesday) { Date.new(2013, 7, 10) }

    it 'returns all workdays when no calendar is selected' do
      expect(spok.workdays).to eq([monday, bovespa_holiday, wednesday])
    end

    it 'returns only bovespa workdays when bovespa calendar is selected' do
      expect(subject.workdays(:bovespa)).to eq([monday, wednesday])
    end
  end

  describe '#to_calendar' do
    context 'when bovespa is the calendar' do
      subject { described_class.new(sunday, tuesday).to_calendar(:bovespa) }
      context 'and start date on a bovespa holiday' do
        let(:start_date) { Date.new(2013, 7, 8) }
        let(:sunday)      { Date.new(2013, 7, 9) } # bovespa holiday
        let(:tuesday)      { Date.new(2013, 7, 11) }
        it 'returns the last bovespa workday as the start date' do
          expect(subject.start_date).to eq(start_date)
        end
      end

      context 'and end date on a bovespa holiday' do
        let(:sunday)    { Date.new(2013, 7, 7) }
        let(:tuesday)    { Date.new(2013, 7, 9) } # bovespa holiday
        let(:end_date) { Date.new(2013, 7, 8) }
        it 'returns the last bovespa workday as the end date' do
          expect(subject.end_date).to eq(end_date)
        end
      end

      context 'and start and end date on a bovespa holiday' do
        let(:sunday)    { Date.new(2012, 12, 31) } #bovespa holiday
        let(:tuesday)    { Date.new(2013, 1, 1) } # bovespa holiday
        it 'start and end date are going to be the same' do
          expect(subject.start_date).to eq(subject.end_date)
        end
      end
    end

    context 'when brasil is the calendar' do
      subject { described_class.new(sunday, tuesday).to_calendar(:brasil) }

      context 'and the start date is holiday' do
        let(:start_date) { Date.new(2013, 9, 6) }
        let(:sunday)      { Date.new(2013, 9, 7) } # brasil holiday
        let(:tuesday)      { Date.new(2013, 9, 10) }
        it 'returns the last workday as the start date' do
          expect(subject.start_date).to eq(start_date)
        end
      end

      context 'end date on a bovespa holiday' do
        let(:sunday)    { Date.new(2013, 9, 5) }
        let(:tuesday)    { Date.new(2013, 9, 7) } # brasil holiday
        let(:end_date) { Date.new(2013, 9, 6) }
        it 'returns the last bovespa workday as the end date' do
          expect(subject.end_date).to eq(end_date)
        end
      end
    end
  end

  describe '#one_day?' do
    context 'when the start and end date are the same' do
      let(:tuesday)    { sunday }
      it { expect(subject).to be_one_day }
    end

    context 'when the start and end date are different' do
      it { expect(subject).not_to be_one_day }
    end
  end

  describe '#days_count' do
    it 'counts number of days on spok' do
      spok = described_class.new(sunday, tuesday)

      expect(spok.days_count).to eq(2)
    end

    it 'returns 0 when start date is equal to the end date' do
      spok = described_class.new(sunday, sunday)

      expect(spok.days_count).to eq(0)
    end
  end

  describe '#years_count' do
    let(:start_date) { Date.new(2013, 1, 1) }

    it 'calculates integer result' do
      end_date = Date.new(2014, 1, 1)

      expect(described_class.new(start_date, end_date).years_count).to eq(1)
    end

    it 'calculates fraction result' do
      spok = described_class.new(start_date, Date.new(2014, 7, 1))
      greater_spok = described_class.new(start_date, Date.new(2014, 9, 1))

      expect(spok.years_count).to eq(1.5)
      expect(greater_spok.years_count).to eq(1.7)
    end
  end

  describe '#==' do
    let(:spok) { Spok.new(sunday, tuesday) }

    context 'when spoks are equal' do
      let(:other_spok) { Spok.new(sunday, tuesday) }

      it { expect(spok == other_spok).to eq(true) }
    end

    context 'when spoks are not equal' do
      let(:other_spok) { Spok.new(sunday, monday) }

      it { expect(spok == other_spok).to eq(false) }
    end

    context 'when the comparison does respond to start date' do
      let(:other_spok) { double(:other_spok, :end_date => double(:end_date)) }
      it { expect(spok == other_spok).to eq(false) }
    end

    context 'when the comparison does respond to end date' do
      let(:other_spok) { double(:other_spok, :start_date => sunday) }
      it { expect(spok == other_spok).to eq(false) }
    end
  end

  describe '#to_range' do
    subject { Spok.new(sunday, tuesday).to_range }
    it { is_expected.to eq((sunday..tuesday)) }
  end
end


spec/lib/period_spec.rb
require 'spec_helper'
require 'period'

describe Period do
  let(:sunday) { Date.new(2012, 1, 1) }
  let(:monday) { Date.new(2012, 1, 2) }
  let(:tuesday) { Date.new(2012, 1, 3) }

  subject(:period) { described_class.new(sunday, tuesday) }

  describe '.parse' do
    it { expect(described_class.parse('20120101-20120103')).to eq(period) }

    context 'when the input string is wrong' do
      it { expect(described_class.parse('20120101')).to be_nil }
      it { expect(described_class.parse(nil)).to be_nil }
      it { expect(described_class.parse('')).to be_nil }
    end
  end

  describe '#date_as_integer' do
    it 'represents the start date as integer' do
      expect(subject.start_date_as_integer).to eq(20120101)
    end

    it 'represents the end date as integer' do
      expect(subject.end_date_as_integer).to eq(20120103)
    end
  end

  describe '#date_as_string' do
    it 'represents the start date as string' do
      expect(subject.start_date_as_string).to eq('20120101')
    end

    it 'represents the end date as string' do
      expect(subject.end_date_as_string).to eq('20120103')
    end
  end

  describe '#validate_period!' do
    it 'does not raise error when start date is before or equals to the end date' do
      expect { subject }.not_to raise_error
      expect { described_class.new(sunday, sunday) }.not_to raise_error
    end

    it 'raises exception when the start date is nil' do
      expect { described_class.new(nil, sunday) }.to raise_error(ArgumentError)
    end

    it 'raises exception when the end date is nil' do
      expect { described_class.new(tuesday, nil) }.to raise_error(ArgumentError)
    end

    it 'raises exception when the start date is greater than the end date' do
      expect { described_class.new(tuesday, sunday) }.to raise_error(ArgumentError)
    end
  end

  describe '#to_s' do
    it 'successfully converts Period to String' do
      expect(period.to_s).to eq('20120101-20120103')
    end
  end

  describe '#workdays' do
    subject(:period) { described_class.new(sunday, wednesday) }

    let(:sunday) { Date.new(2013, 7, 7) }
    let(:monday) { Date.new(2013, 7, 8) }
    let(:bovespa_holiday) { Date.new(2013, 7, 9) }
    let(:wednesday) { Date.new(2013, 7, 10) }

    it 'returns all workdays when no calendar is selected' do
      expect(period.workdays).to eq([monday, bovespa_holiday, wednesday])
    end

    it 'returns only bovespa workdays when bovespa calendar is selected' do
      expect(subject.workdays(:bovespa)).to eq([monday, wednesday])
    end
  end

  describe '#to_calendar' do
    context 'when bovespa is the calendar' do
      subject { described_class.new(sunday, tuesday).to_calendar(:bovespa) }
      context 'and start date on a bovespa holiday' do
        let(:start_date) { Date.new(2013, 7, 8) }
        let(:sunday)      { Date.new(2013, 7, 9) } # bovespa holiday
        let(:tuesday)      { Date.new(2013, 7, 11) }
        it 'returns the last bovespa workday as the start date' do
          expect(subject.start_date).to eq(start_date)
        end
      end

      context 'and end date on a bovespa holiday' do
        let(:sunday)    { Date.new(2013, 7, 7) }
        let(:tuesday)    { Date.new(2013, 7, 9) } # bovespa holiday
        let(:end_date) { Date.new(2013, 7, 8) }
        it 'returns the last bovespa workday as the end date' do
          expect(subject.end_date).to eq(end_date)
        end
      end

      context 'and start and end date on a bovespa holiday' do
        let(:sunday)    { Date.new(2012, 12, 31) } #bovespa holiday
        let(:tuesday)    { Date.new(2013, 1, 1) } # bovespa holiday
        it 'start and end date are going to be the same' do
          expect(subject.start_date).to eq(subject.end_date)
        end
      end
    end

    context 'when brasil is the calendar' do
      subject { described_class.new(sunday, tuesday).to_calendar(:brasil) }

      context 'and the start date is holiday' do
        let(:start_date) { Date.new(2013, 9, 6) }
        let(:sunday)      { Date.new(2013, 9, 7) } # brasil holiday
        let(:tuesday)      { Date.new(2013, 9, 10) }
        it 'returns the last workday as the start date' do
          expect(subject.start_date).to eq(start_date)
        end
      end

      context 'end date on a bovespa holiday' do
        let(:sunday)    { Date.new(2013, 9, 5) }
        let(:tuesday)    { Date.new(2013, 9, 7) } # brasil holiday
        let(:end_date) { Date.new(2013, 9, 6) }
        it 'returns the last bovespa workday as the end date' do
          expect(subject.end_date).to eq(end_date)
        end
      end
    end
  end

  describe '#one_day_period?' do
    context 'when the start and end date are the same' do
      let(:tuesday)    { sunday }
      it { expect(subject).to be_one_day_period }
    end

    context 'when the start and end date are different' do
      it { expect(subject).not_to be_one_day_period }
    end
  end

  describe '#days_count' do
    it 'counts number of days on period' do
      period = described_class.new(sunday, tuesday)

      expect(period.days_count).to eq(2)
    end

    it 'returns 0 when start date is equal to the end date' do
      period = described_class.new(sunday, sunday)

      expect(period.days_count).to eq(0)
    end
  end

  describe '#years_count' do
    let(:start_date) { Date.new(2013, 1, 1) }

    it 'calculates integer result' do
      end_date = Date.new(2014, 1, 1)

      expect(described_class.new(start_date, end_date).years_count).to eq(1)
    end

    it 'calculates fraction result' do
      period = described_class.new(start_date, Date.new(2014, 7, 1))
      greater_period = described_class.new(start_date, Date.new(2014, 9, 1))

      expect(period.years_count).to eq(1.5)
      expect(greater_period.years_count).to eq(1.7)
    end
  end

  describe '#==' do
    let(:period) { Period.new(sunday, tuesday) }

    context 'when periods are equal' do
      let(:other_period) { Period.new(sunday, tuesday) }

      it { expect(period == other_period).to eq(true) }
    end

    context 'when periods are not equal' do
      let(:other_period) { Period.new(sunday, monday) }

      it { expect(period == other_period).to eq(false) }
    end

    context 'when the comparison does respond to start date' do
      let(:other_period) { double(:other_period, :end_date => double(:end_date)) }
      it { expect(period == other_period).to eq(false) }
    end

    context 'when the comparison does respond to end date' do
      let(:other_period) { double(:other_period, :start_date => sunday) }
      it { expect(period == other_period).to eq(false) }
    end
  end

  describe '#to_range' do
    subject { Period.new(sunday, tuesday).to_range }
    it { is_expected.to eq((sunday..tuesday)) }
  end
end


lib/spok.rb
require "spok/version"
require 'spok/workday'

# Public: Class responsible for dealing with periods of Dates, considering
# workdays and restdays.
class Spok
  # Internal: String specifying format for dates in the period.
  DATE_FORMAT = '%Y%m%d'
  attr_reader :start_date, :end_date

  # Public: Parses a string into a Spok.
  #
  # dates_string - String containing the start and end dates for a period of
  #                 days separated by a dash.
  #
  # Examples
  #
  #   Spok.parse('20120101-20120103')
  #   # => #<Spok:0x00007f951e8a2ea0 ...>
  #
  #   Spok.parse('invalid string')
  #   # => nil
  #
  # Returns a Spok or nil when the string does not contain two valid dates.
  def self.parse(dates_string)
    return nil unless dates_string

    start_date, end_date = dates_string.split('-')

    if start_date && end_date
      Spok.new(::Date.parse(start_date), ::Date.parse(end_date))
    else
      nil
    end
  end

  # Public: Initialize a Spok.
  #
  # start_date - Initial Date for the Spok.
  # end_date - Final Date for Spok.
  def initialize(start_date, end_date)
    @start_date = start_date
    @end_date = end_date
    validate!
  end

  # Public: Returns the Spok start date as an Integer.
  #
  # Examples
  #
  #   spok.start_date_as_integer
  #   # => 20120101
  def start_date_as_integer
    as_integer @start_date
  end

  # Public: Returns the Spok end date as an Integer.
  #
  # Examples
  #
  #   spok.end_date_as_integer
  #   # => 20120103
  def end_date_as_integer
    as_integer @end_date
  end

  # Public: Returns the Spok start date as a String.
  #
  # Examples
  #
  #   spok.start_date_as_string
  #   # => "20120101"
  def start_date_as_string
    as_string @start_date
  end

  # Public: Returns the Spok end date as a String.
  #
  # Examples
  #
  #   spok.end_date_as_string
  #   # => "20120103"
  def end_date_as_string
    as_string @end_date
  end

  # Public: Returns an array containing all workdays on Spok.
  #
  # calendar - Symbol informing in which calendar to check for workdays
  #            (default: :brasil).
  #
  # Examples
  #
  #   spok.workdays
  #   # => [Mon, 02 Jan 2012, Tue, 03 Jan 2012]
  def workdays(calendar = :brasil)
    (@start_date..@end_date).to_a.delete_if{ |date| Workday.restday?(date, calendar: calendar) }
  end

  # Public: Returns a Spok containing the same dates in a different calendar.
  #
  # calendar - Symbol informing calendar for new Spok (default: :bovespa).
  #
  # Examples
  #
  #   spok.to_calendar(:bovespa)
  #   # => #<Spok:0x00007fbf122dba08 ...>
  def to_calendar(calendar = :bovespa)
    Spok.new(
      Workday.last_workday(@start_date, calendar: calendar),
      Workday.last_workday(@end_date, calendar: calendar)
    )
  end

  # Public: Returns an array containing all Dates on Spok.
  #
  # Examples
  #
  #   spok.to_a
  #   # => [Sun, 01 Jan 2012, Mon, 02 Jan 2012, Tue, 03 Jan 2012]
  def to_a
    (@start_date..@end_date).to_a
  end

  # Public: Returns a String containing the Spok start and end date separated
  # by a dash.
  #
  # Examples
  #
  #   spok.to_s
  #   # => "20120101-20120103"
  def to_s
    "#{start_date_as_string}-#{end_date_as_string}"
  end

  # Public: Informs whether the Spok has just one day or not.
  #
  # Examples
  #
  #   spok.one_day?
  #   # => false
  #
  # Returns a boolean.
  def one_day?
    @start_date == @end_date
  end

  # Public: Returns an Integer with the number of days in the Spok.
  #
  # Examples
  #
  #   spok.days_count
  #   # => 2
  def days_count
    (@end_date - @start_date).to_i
  end

  # Public: Returns a Float with the number of years in the Spok.
  #
  # Examples
  #
  #   spok.years_count
  #   # => 1.6
  def years_count
    ((@end_date - @start_date).to_f / 365).to_f.round(1)
  end

  # Public: Compares the Spok with other Spok. Two spoks are considered
  # equal when they are both instances of Spok, and have the same start and
  # end dates.
  #
  # other_spok - Spok to be checked against.
  #
  # Examples
  #
  #   spok == other_spok
  #   # => false
  #
  # Returns a boolean.
  def ==(other_spok)
    other_spok.class == self.class &&
    other_spok.start_date == @start_date &&
    other_spok.end_date == @end_date
  end

  # Public: Returns a range containing the Dates in the Spok.
  #
  # Examples
  #
  #   spok.to_range
  #   # => Sun, 01 Jan 2012..Tue, 03 Jan 2012
  def to_range
    (@start_date..@end_date)
  end

  private

  def as_string(date)
    date.strftime(DATE_FORMAT)
  end

  def as_integer(date)
    date.strftime(DATE_FORMAT).to_i
  end

  def validate!
    raise ArgumentError.new("Start date must be present.") unless @start_date
    raise ArgumentError.new("End date must be present.") unless @end_date
    if @start_date > @end_date
      raise ArgumentError.new("End date (#{@end_date}) must be greater or equal to start date (#{@start_date})")
    end
  end
end


lib/spok/workday.rb
require 'active_support'
require 'active_support/core_ext'
require 'set'
require 'yaml'

class Spok
  # Public: Various methods useful for checking for restdays and workdays.
  # All methods are module methods and should be called on the Spok::Workday
  # module.
  module Workday
    # Public: Array of available calendars.
    CALENDARS = %i(brasil bovespa)

    # Public: Hash containing all holidays for each available calendar.
    HOLIDAYS = CALENDARS.map do |calendar|
      holidays_file = File.open(File.join(File.dirname(__FILE__), "config/#{calendar}.yml"))
      holidays = YAML.safe_load(holidays_file.read, [Date])
      [calendar, Set.new(holidays[calendar.to_s])]
    end.to_h

    # Public: Checks if a given day is a restday.
    #
    # date     - The Date to be checked.
    # calendar - Symbol informing in which calendar the date will be checked
    #            (default: :brasil).
    #
    # Examples
    #
    #   Spok::Workday.restday?(Date.new(2012, 8, 6))
    #   # => false
    #
    # Returns a boolean.
    def self.restday?(date, calendar: :brasil)
      self.weekend?(date) || self.holiday?(date, calendar: calendar)
    end

    # Public: Checks if a given day is a workday.
    #
    # date     - The Date to be checked.
    # calendar - Symbol informing in which calendar the date will be checked
    #            (default: :brasil).
    #
    # Examples
    #
    #   Spok::Workday.workday?(Date.new(2012, 8, 6))
    #   # => true
    #
    # Returns a boolean.
    def self.workday?(date, calendar: :brasil)
      !restday?(date, calendar: calendar)
    end

    # Public: Checks if a given Date is on a weekend.
    #
    # date - The Date to be checked.
    #
    # Examples
    #
    #   Spok::Workday.weekend?(Date.new(2012, 8, 6))
    #   # => false
    #
    # Returns a boolean.
    def self.weekend?(date)
      weekday = date.wday

      weekday == 0 || weekday == 6
    end

    # Public: Checks if a given Date is on a holiday.
    #
    # date     - The Date to be checked.
    # calendar - Symbol informing in which calendar the date will be checked
    #            (default: :brasil).
    #
    # Examples
    #
    #   Spok::Workday.holiday?(Date.new(2012, 5, 1))
    #   # => true
    #
    # Returns a boolean.
    def self.holiday?(date, calendar: :brasil)
      HOLIDAYS[calendar].include?(date.to_date)
    end

    # Public: Returns the last workday until the informed date.
    # It returns the informed date in case it is a workday.
    #
    # date     - End Date to check for workdays.
    # calendar - Symbol informing in which calendar to check for workdays
    #            (default: :brasil).
    #
    # Examples
    #   Spok::Workday.last_workday(Date.new(2012, 10, 21))
    #   # => #<Date: 2012-10-19 ((2456220j,0s,0n),+0s,2299161j)>
    def self.last_workday(date, calendar: :brasil)
      return date if workday?(date, calendar: calendar)

      last_workday((date - 1.day), calendar: calendar)
    end

    # Public: Returns the next workday starting from the informed date.
    # It returns the informed date in case it is a workday.
    #
    # date     - Start Date to check for workdays.
    # calendar - Symbol informing in which calendar to check for workdays
    #            (default: :brasil).
    #
    # Examples
    #   Spok::Workday.next_workday(Date.new(2012, 10, 21))
    #   # => #<Date: 2012-10-19 ((2456220j,0s,0n),+0s,2299161j)>
    def self.next_workday(date, calendar: :brasil)
      return date if workday?(date, calendar: calendar)

      next_workday((date + 1.day), calendar: calendar)
    end
  end
end


lib/period.rb
require "period/version"
require 'period/workday'

# Public: Class responsible for dealing with periods of Dates, considering
# workdays and restdays.
class Period
  # Internal: String specifying format for dates in the period.
  DATE_FORMAT = '%Y%m%d'
  attr_reader :start_date, :end_date

  # Public: Parses a string into a Period.
  #
  # period_string - String containing the period start and end date separated
  #                 by a dash.
  #
  # Examples
  #
  #   Period.parse('20120101-20120103')
  #   # => #<Period:0x00007f951e8a2ea0 ...>
  #
  #   Period.parse('invalid string')
  #   # => nil
  #
  # Returns a Period or nil when the string does not contain two valid dates.
  def self.parse(period_string)
    return nil unless period_string

    start_date, end_date = period_string.split('-')

    if start_date && end_date
      Period.new(::Date.parse(start_date), ::Date.parse(end_date))
    else
      nil
    end
  end

  # Public: Initialize a Period.
  #
  # start_date - Initial Date for the Period.
  # end_date - Final Date for Period.
  def initialize(start_date, end_date)
    @start_date = start_date
    @end_date = end_date
    validate_period!
  end

  # Public: Returns the Period start date as an Integer.
  #
  # Examples
  #
  #   period.start_date_as_integer
  #   # => 20120101
  def start_date_as_integer
    as_integer @start_date
  end

  # Public: Returns the Period end date as an Integer.
  #
  # Examples
  #
  #   period.end_date_as_integer
  #   # => 20120103
  def end_date_as_integer
    as_integer @end_date
  end

  # Public: Returns the Period start date as a String.
  #
  # Examples
  #
  #   period.start_date_as_string
  #   # => "20120101"
  def start_date_as_string
    as_string @start_date
  end

  # Public: Returns the Period end date as a String.
  #
  # Examples
  #
  #   period.end_date_as_string
  #   # => "20120103"
  def end_date_as_string
    as_string @end_date
  end

  # Public: Returns an array containing all workdays on Period.
  #
  # calendar - Symbol informing in which calendar to check for workdays
  #            (default: :brasil).
  #
  # Examples
  #
  #   period.workdays
  #   # => [Mon, 02 Jan 2012, Tue, 03 Jan 2012]
  def workdays(calendar = :brasil)
    (@start_date..@end_date).to_a.delete_if{ |date| Workday.restday?(date, calendar: calendar) }
  end

  # Public: Returns a Period containing the same dates in a different calendar.
  #
  # calendar - Symbol informing calendar for new Period (default: :bovespa).
  #
  # Examples
  #
  #   period.to_calendar(:bovespa)
  #   # => #<Period:0x00007fbf122dba08 ...>
  def to_calendar(calendar = :bovespa)
    Period.new(
      Workday.last_workday(@start_date, calendar: calendar),
      Workday.last_workday(@end_date, calendar: calendar)
    )
  end

  # Public: Returns an array containing all Dates on Period.
  #
  # Examples
  #
  #   period.to_a
  #   # => [Sun, 01 Jan 2012, Mon, 02 Jan 2012, Tue, 03 Jan 2012]
  def to_a
    (@start_date..@end_date).to_a
  end

  # Public: Returns a String containing the Period start and end date separated
  # by a dash.
  #
  # Examples
  #
  #   period.to_s
  #   # => "20120101-20120103"
  def to_s
    "#{start_date_as_string}-#{end_date_as_string}"
  end

  # Public: Informs whether the Period has just one day or not.
  #
  # Examples
  #
  #   period.one_day_period?
  #   # => false
  #
  # Returns a boolean.
  def one_day_period?
    @start_date == @end_date
  end

  # Public: Returns an Integer with the number of days in the Period.
  #
  # Examples
  #
  #   period.days_count
  #   # => 2
  def days_count
    (@end_date - @start_date).to_i
  end

  # Public: Returns a Float with the number of years in the Period.
  #
  # Examples
  #
  #   period.years_count
  #   # => 1.6
  def years_count
    ((@end_date - @start_date).to_f / 365).to_f.round(1)
  end

  # Public: Compares the Period with other Period. Two periods are considered
  # equal when they are both instances of Period, and have the same start and
  # end dates.
  #
  # other_period - Period to be checked against.
  #
  # Examples
  #
  #   period == other_period
  #   # => false
  #
  # Returns a boolean.
  def ==(other_period)
    other_period.class == self.class &&
    other_period.start_date == @start_date &&
    other_period.end_date == @end_date
  end

  # Public: Returns a range containing the Dates in the Period.
  #
  # Examples
  #
  #   period.to_range
  #   # => Sun, 01 Jan 2012..Tue, 03 Jan 2012
  def to_range
    (@start_date..@end_date)
  end

  private

  def as_string(date)
    date.strftime(DATE_FORMAT)
  end

  def as_integer(date)
    date.strftime(DATE_FORMAT).to_i
  end

  def validate_period!
    raise ArgumentError.new("Start date must be present.") unless @start_date
    raise ArgumentError.new("End date must be present.") unless @end_date
    if @start_date > @end_date
      raise ArgumentError.new("End date (#{@end_date}) must be greater or equal to start date (#{@start_date})")
    end
  end
end
